### PR TITLE
bpo-42545: Check that all symbols in the limited ABI are exported

### PR DESCRIPTION
**Pull request by pablogsal**
https://bugs.python.org/issue42545

### CODE REVIEW

File: Tools/scripts/stable_abi.py [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+        )
+        return 1
+    return 0
+
+
+def generate_limited_api_symbols(args):
+    library = sysconfig.get_config_var("LIBRARY")
+    ldlibrary = sysconfig.get_config_var("LDLIBRARY")
+    if ldlibrary != library:
+        raise Exception("Limited ABI symbols can only be generated from a static build")
```
Comment (vstinner) [inline Tools/scripts/stable_abi.py:83]: What is the issue with the libpython dynamic library?

Recently, I modified Tools/scripts/smelly.py to check not only the static library, but also the libpython dynamic library and C extensions.

**Review by vstinner — COMMENTED** [review]
Nice script!

Should we check if PC/python3dll.c is outdated?

**Comment by pablogsal** [comment]
> Should we check if PC/python3dll.c is outdated?

I considered that, but that file has some other functions being exposed that are Windows dependent so we cannot technically use this script that runs on Linux. I will try to figure out another PR to keep that file updated.

File: Tools/scripts/stable_abi.py (hunk 1 above)
Comment (pablogsal) [reply Tools/scripts/stable_abi.py:83]: > What is the issue with the libpython dynamic library?

That the dynamic library may have some symbols missing if is built-in weird ways. Honestly, the reason I am raising here was to keep the script as simple as possible on the first version :)

**Comment by vstinner** [comment]
> I considered that, but that file has some other functions being exposed that are Windows dependent so we cannot technically use this script that runs on Linux.

My request was more about checking from your script output if python3dll.c is complete or not. If there are differences between Windows and Linux in the ABI, maybe differences can be saved in a file used by your script?

> I will try to figure out another PR to keep that file updated.

Sure. It's ok to first only focus on Linux.

**Comment by vstinner** [comment]
The CI fails :-) 

```
Some symbols from the stable ABI are missing: PyCFunction_New
Makefile:1922: recipe for target 'check-limited-abi' failed
```

File: Tools/scripts/stable_abi.py (hunk 1 above)
Comment (vstinner) [reply Tools/scripts/stable_abi.py:83]: Ok.

**Comment by pablogsal** [comment]
> The CI fails :-)

Whoops, wrong makefile invocation for shared libs. Fixed!

**Comment by pablogsal** [comment]
> If there are differences between Windows and Linux in the ABI, maybe differences can be saved in a file used by your script?

I checked and the file needs some love (functions missing or old functions remaining), so I am going to defer fixes to that file for a later PR, maybe checking with Steve.

**Review by tryalice — COMMENTED** [review]
[README.txt](https://github.com/python/cpython/files/5646063/README.txt)

**Comment by pablogsal** [comment]
> [README.txt](https://github.com/python/cpython/files/5646063/README.txt)

?

**Pull request merged**